### PR TITLE
Feature 2085 convert R script (pntnc2ascii.r) to python

### LINKED ISSUE #2085: Reimplement the pntnc2ascii.R utility Rscript in Python.
## Describe the Enhancement ##
This issue arose via this [METplus Discussion](https://github.com/dtcenter/METplus/discussions/1472). The METplus team recommended that @PerryShafran-NOAA run the pntnc2ascii.R Rscript to dump the NetCDF output from PB2NC prior to interrogating that data further. That script is very useful and really helps with debugging, but it cannot be run on WCOSS or Hera due to R package dependencies.

This task is to reimplement this functionality in Python instead of R, making sure that it uses Python packages actually available on WCOSS and Hera.

The Rscript currently lives in "met/scripts/Rscripts". The python implementation could be stored in "met/scripts/python". However, I'd recommend re-evaluating the final location for this. Does it really belong in the METplus repo? Or perhaps the METdatadb repo after it's been named to METdataio? Or perhaps it does belong in MET, but should be placed in a new "met/scripts/utility" directory?

Recommend removing the old pntnc2ascii.R script once the Python version has replaced it. 

Note that the majority of this work is likely already done in [read_met_point_obs.py](https://github.com/dtcenter/MET/blob/develop/met/scripts/python/read_met_point_obs.py). We could just modify it to provide an option to write the observations to stdout. Or support a "-out" command line option to write the data to an output file.

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
No sub-issues needed.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
2773542

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
No impacts.

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by hsoh-u**
## Expected Differences ##

Note: There is a warning with older netCDF4 package.
- warning with netCDF4 1.5.3
- no warning with netCDF4 1.5.6
```
print_nc2ascii.py:78: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
  self.use_var_id = nc_group.use_var_id.lower() == "true"
```

Two options were added
- --add-header: the header is added (default, no header)
- --use-comma: allows to generate comma separated text file (default: disabled)

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [ ] Describe testing already performed for these changes:</br>

python3 /d1/personal/hsoh/git/features/feature_2085_R_to_python_pntnc2ascii_R/MET/scripts/python/print_nc2ascii.py /d1/projects/MET/MET_regression/develop/NB20221110/MET-develop/test_output/ascii2nc/edr_hourly.20130827.nc | less

python3 /d1/personal/hsoh/git/features/feature_2085_R_to_python_pntnc2ascii_R/MET/scripts/python/print_nc2ascii.py /d1/projects/MET/MET_regression/develop/NB20221110/MET-develop/test_output/ascii2nc/edr_hourly.20130827.nc | less

python3 /d1/personal/hsoh/git/features/feature_2085_R_to_python_pntnc2ascii_R/MET/scripts/python/print_nc2ascii.py /d1/projects/MET/MET_regression/develop/NB20221110/MET-develop/test_output/pb2nc/ndas.20120409.t12z.prepbufr.tm00.nc | less

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[No]**

- [x] Do these changes include sufficient testing updates? **[No]**

This is a standalone utility

- [x] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Development** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by JohnHalleyGotway** [comment]
@DanielAdriaansen, I'm wondering if you're planning to review this PR from Howard?

I see we're both listed but perhaps we're both thinking the other is going to review it ;)

**Comment by DanielAdriaansen** [comment]
I have a few comments and questions:
1) I ran the Rscript and compared with the Python script and noticed that the precision printed is different. From the e-mail chain "Questions for printing out MET point observation as text by python script", I thought we had agreed to align them but maybe not. For example:
Rscript:
`MEDEDR N571UA 20130827_130000 31.9179  -96.2782 10973.41 200 NA 36002.00 1.00 0.05`
Python:
`MEDEDR N571UA 20130827_130000  31.9179 -96.2782 10973.4102 200  NA 36002.0000 1.00 0.05`

The "elv" and "height" columns look like they need to have some formatting applied to the print statement. Could you add formatting to these columns @hsoh-u?

2) In the issue, @JohnHalleyGotway says:
> However, I'd recommend re-evaluating the final location for this. Does it really belong in the METplus repo? Or perhaps the METdatadb repo after it's been named to METdataio? Or perhaps it does belong in MET, but should be placed in a new "met/scripts/utility" directory?

Do we need to change the location of the new `print_nc2ascii.py` script?

3) I recommend renaming the script from `print_nc2ascii.py` to `print_pointnc2ascii.py`.

4) In the issue description, @JohnHalleyGotway mentions:
> Note that the majority of this work is likely already done in read_met_point_obs.py.

What is the difference between `read_met_point_obs.py` and `print_nc2ascii.py`?

5) Is there any current documentation about `pntnc2ascii.R`? If there is not, should we add documentation about this highly useful utility script somewhere?

6) As a final step, we should remove the `pntnc2ascii.R` script, as noted in the issue.

**Comment by hsoh-u** [comment]
> I have a few comments and questions:
> 
>     1. I ran the Rscript and compared with the Python script and noticed that the precision printed is different. From the e-mail chain "Questions for printing out MET point observation as text by python script", I thought we had agreed to align them but maybe not. For example:
>        Rscript:
>        `MEDEDR N571UA 20130827_130000 31.9179  -96.2782 10973.41 200 NA 36002.00 1.00 0.05`
>        Python:
>        `MEDEDR N571UA 20130827_130000  31.9179 -96.2782 10973.4102 200  NA 36002.0000 1.00 0.05`
> 
> 
> The "elv" and "height" columns look like they need to have some formatting applied to the print statement. Could you add formatting to these columns @hsoh-u?
> 
>     2. In the issue, @JohnHalleyGotway says:
> 
> 
> > However, I'd recommend re-evaluating the final location for this. Does it really belong in the METplus repo? Or perhaps the METdatadb repo after it's been named to METdataio? Or perhaps it does belong in MET, but should be placed in a new "met/scripts/utility" directory?
> 
> Do we need to change the location of the new `print_nc2ascii.py` script?
> 
>     3. I recommend renaming the script from `print_nc2ascii.py` to `print_pointnc2ascii.py`.
> 
>     4. In the issue description, @JohnHalleyGotway mentions:
> 
> 
> > Note that the majority of this work is likely already done in read_met_point_obs.py.
> 
> What is the difference between `read_met_point_obs.py` and `print_nc2ascii.py`?
> 
>     5. Is there any current documentation about `pntnc2ascii.R`? If there is not, should we add documentation about this highly useful utility script somewhere?
> 
>     6. As a final step, we should remove the `pntnc2ascii.R` script, as noted in the issue.

- The formatting was not enhanced for this release. It will be implemented later.
- I will rename it and move to met/scripts/utility
- read_met_point_obs.py is an example of the customized python embedding for the point observation. The base class of the read_met_point_obs.py is the met_point_obs.py. met_point_obs.py has capability to read the MET's point observation NetCDF for the testing purpose. Two cases - one as the MET point obs NetCDF and other which reads the same data by the python script - process the same input. They must produce the same output for MET tools. Some codes at met_point_obs.py are copied to print_nc2ascii.py.

**Comment by hsoh-u** [comment]
- Renamed and moved scripts/utility/print_pointnc2ascii.py
  - will be installed at MET_BASE/shared/met/utility/print_pointnc2ascii.py
- The columns are aligned
- -o "out_file_name" (or --out="out_file_name") option is added

File: scripts/utility/print_pointnc2ascii.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+from optparse import OptionParser
+
+import numpy as np
+from netCDF4 import Dataset
+
+def usage():
+    print(f'Usage: python3 print_nc2ascii.py MET_point_obs_nc <--hide-header> <--use-comma>')
+    print(f'            MET_point_obs_nc: NetCDF filename to read (required)')
+    print(f'                --add-header: to add the header (optional, default: not included)')
+    print(f'                 --use-comma: use the "comma" as separator for comma separated output (optional, default: use spaces)"')
```
Comment (DanielAdriaansen) [inline scripts/utility/print_pointnc2ascii.py:23]: @hsoh-u can you update the `usage()` here to reflect the option for the outfile options? Thanks.

**Review by DanielAdriaansen — CHANGES_REQUESTED** [review]
Just a minor request.

File: scripts/utility/print_pointnc2ascii.py (hunk 1 above)
Comment (hsoh-u) [reply scripts/utility/print_pointnc2ascii.py:23]: Thank you for catching this. It's updated.

The python way to get the help message:
```
python3 print_pointnc2ascii.py -h
```

Comment (DanielAdriaansen) [reply scripts/utility/print_pointnc2ascii.py:23]: Thanks Howard. I think the "-h" is optional, is that right? So:
```
python3 print_pointnc2ascii.py -h
```

is the same as:
```
python3 print_pointnc2ascii.py
```
?

Comment (hsoh-u) [reply scripts/utility/print_pointnc2ascii.py:23]: No, they are different
utility/print_pointnc2ascii.py  -h
```
Usage: print_pointnc2ascii.py [options]

Options:
  -h, --help            show this help message and exit
  --add-header, --add_header
                         Add header (default: not include header)
  --use-comma, --use_comma
                         Use comma as separator (default: False)
  -o OUT_FILE, --out=OUT_FILE
                         Save the text into the file (default: False)
```

utility/print_pointnc2ascii.py
```
 == INFO == Missing input NetCDF filename
Usage: python3 print_nc2ascii.py MET_point_obs_nc <--hide-header> <--use-comma> <--out=out_filename>
            MET_point_obs_nc: NetCDF filename to read (required)
                --add-header: to add the header (optional, default: not included)
                 --use-comma: use the "comma" as separator for comma separated output (optional, default: use spaces)"
          --out=out_filename: save the text into the file (optional, default: display to the screen)"
       Note: <> indicates optional arguments
```

Comment (DanielAdriaansen) [reply scripts/utility/print_pointnc2ascii.py:23]: Nevermind, I see the difference. I will recommend "-h" in the docs.

**Comment by DanielAdriaansen** [comment]
@JohnHalleyGotway you can view the documentation I added here: https://met.readthedocs.io/en/feature_2085_r_to_python_pntnc2ascii_r/Users_Guide/reformat_point.html#point-netcdf-to-ascii-python-utility

What do you think?

**Review by DanielAdriaansen — APPROVED** [review]
Thanks for doing this Howard, this tool will be a great addition for users that need to inspect the point NetCDF files MET makes. I approve of the changes.

**Pull request merged**